Optimize Intersection Observer computations for inlines

smfr opened the pull request:
#### 32beea64ac2d217858f763d4ac4a877e093b7c13
<pre>
Optimize Intersection Observer computations for inlines
<a href="https://bugs.webkit.org/show_bug.cgi?id=259717">https://bugs.webkit.org/show_bug.cgi?id=259717</a>
rdar://113238475

Reviewed by Ryosuke Niwa.

When computing intersections for RenderInlines, the code computed absolute quads for the RenderInline and
its continutations, then mapped those to local coordinates. What a waste!

Instead we can just compute those rectangles in local coordinates directly. This drops the time spent in
Document::updateIntersectionObservations() from 24% of the time to 14% of the time scrolling an amazon.com
search results page.

* Source/WebCore/page/IntersectionObserver.cpp:
(WebCore::IntersectionObserver::computeIntersectionState const):

Canonical link: <a href="https://commits.webkit.org/266501@main">https://commits.webkit.org/266501@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3755d61f44d4a8b391b4ab3af6fafa42da5723e1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/13988 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/14303 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/14638 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/15726 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/13275 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/14072 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/16812 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/14387 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/15944 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/14158 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/14747 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/11850 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/16434 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/12033 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/12611 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/19639 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/13109 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/12775 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/15982 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/13323 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/11184 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/12598 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3389 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/16932 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/13167 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/page/IntersectionObserver.cpp:
```diff
@@ -365,11 +365,9 @@ auto IntersectionObserver::computeIntersectionState(const IntersectionObserverRe
-            auto absoluteTargetBounds = unitedBoundingBoxes(quads);
-            return enclosingLayoutRect(targetRenderer->absoluteToLocalQuad(absoluteTargetBounds).boundingBox());
+            Vector<LayoutRect> rects;
+            targetRenderer->boundingRects(rects, { });
```
Maybe not in this patch but it seems like it's more natural for the Vector to be the return value instead?

smfr committed:
Optimize Intersection Observer computations for inlines
https://bugs.webkit.org/show_bug.cgi?id=259717
rdar://113238475

Reviewed by Ryosuke Niwa.

When computing intersections for RenderInlines, the code computed absolute quads for the RenderInline and
its continutations, then mapped those to local coordinates. What a waste!

Instead we can just compute those rectangles in local coordinates directly. This drops the time spent in
Document::updateIntersectionObservations() from 24% of the time to 14% of the time scrolling an amazon.com
search results page.

* Source/WebCore/page/IntersectionObserver.cpp:
(WebCore::IntersectionObserver::computeIntersectionState const):

Canonical link: https://commits.webkit.org/266501@main
